File: algorithms/python/numberofDifferentSubsequencesGCDs/numberofDifferentSubsequencesGCDs.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/number-of-different-subsequences-gcds/
# Author: Miao Zhang
# Date:   2021-06-09

class Solution:
    def countDifferentSubsequenceGCDs(self, nums: List[int]) -> int:
        maxnum = max(nums)
        g = [False for _ in range(maxnum + 1)]
        for num in nums: g[num] = True
        res = 0
        for i in range(1, maxnum + 1):
            tmp = 0
            for j in range(i, maxnum + 1, i):
                if g[j] == True:
                    tmp = math.gcd(tmp, j)
            if tmp == i: res += 1
        return res

File: algorithms/cpp/numberofDifferentSubsequencesGCDs/numberofDifferentSubsequencesGCDs.cpp
// Source: https://leetcode.com/problems/number-of-different-subsequences-gcds/
// Author: Miao Zhang
// Date:   2021-06-09

class Solution {
public:
    int countDifferentSubsequenceGCDs(vector<int>& nums) {
        int maxnum = *max_element(begin(nums), end(nums));
        vector<bool> g(maxnum + 1, false);
        for (int num: nums) g[num] = true;
        int res = 0;
        for (int i = 1; i <= maxnum; i++) {
            int gcd = 0;
            for (int j = i; j <= maxnum; j += i) {
                if (g[j] == true) {
                    gcd = __gcd(gcd, j);
                }
            }
            if (gcd == i) res++;
        }
        return res;
    }
};
